samples/bluetooth/bap_broadcast_sink/src/main.c

```diff
@@ -281,83 +289,115 @@ static bool subgroup_get_valid_bis_indexes_cb(const struct bt_bap_base_subgroup
+				/* Exact match */
+				printk("Channel allocation match, exact. BIS idx bitfield 0x%x\n",
 				       bis_indexes);
-				*(uint32_t *)user_data = bis_indexes;
 
-				return false; /* bis indexes found */
+				bis_index->bis_index_bitfield[bis_index->num_subgroups] =
+					bis_indexes;
+				break; /* no need to go futher */
 			}
+
+			/* Partial match */
+			printk("Channel allocation match, partial. BIS idx bitfield 0x%x\n",
+			       bis_indexes);
 		}
 	}
+
+	bis_index->num_subgroups++;
+
 	return true; /* continue to next subgroup */
 }
+#else /* !CONFIG_TARGET_BROADCAST_CHANNEL */
+static bool subgroup_get_bis_indexes_cb(const struct bt_bap_base_subgroup *subgroup,
+					void *user_data)
+{
+	struct bis_index_data *bis_index = user_data;
+	const int err = bt_bap_base_subgroup_get_bis_indexes(
+		subgroup, &bis_index->bis_index_bitfield[bis_index->num_subgroups]);
+
+	if (err != 0) {
+		printk("Failed to parse all BIS: %d", err);
+
+		return false;
+	}
+
+	printk("bis_index.bis_index_bitfield[%d] = 0x%0x\n", bis_index->num_subgroups,
+	       bis_index->bis_index_bitfield[bis_index->num_subgroups]);
+
+	bis_index->num_subgroups++;
+
+	return true;
+}
 #endif /* CONFIG_TARGET_BROADCAST_CHANNEL */
 
 static void base_recv_cb(struct bt_bap_broadcast_sink *sink, const struct bt_bap_base *base,
 			 size_t base_size)
 {
-	uint32_t base_bis_index_bitfield = 0U;
 	int err;
 
+	base_bis_index_bitfield = 0U;
+
 	if (base_received) {
 		return;
 	}
 
 	printk("Received BASE with %d subgroups from broadcast sink %p\n",
 	       bt_bap_base_get_subgroup_count(base), sink);
 
+	memset(&bis_index, 0, sizeof(struct bis_index_data));
+
 #if defined(CONFIG_TARGET_BROADCAST_CHANNEL)
 	/**
 	 * Get a 32-bit bitfield of BIS indexes that cover the channel allocation values in
 	 * CONFIG_TARGET_BROADCAST_CHANNEL.
 	 */
 	printk("Target channel location: 0x%x\n", CONFIG_TARGET_BROADCAST_CHANNEL);
-	err = bt_bap_base_foreach_subgroup(base, subgroup_get_valid_bis_indexes_cb,
-					   &base_bis_index_bitfield);
-	if ((err != 0 && err != -ECANCELED) ||
-	    (err == -ECANCELED && base_bis_index_bitfield == 0)) {
-		printk("Failed to get valid BIS indexes: %d\n", err);
-
-		return;
-	}
+	err = bt_bap_base_foreach_subgroup(base, subgroup_get_valid_bis_indexes_cb, &bis_index);
 #else
-	err = bt_bap_base_get_bis_indexes(base, &base_bis_index_bitfield);
-	if (err != 0) {
-		printk("Failed to get BIS indexes: %d\n", err);
+	/* Get BIS index data for each subgroup */
+	err = bt_bap_base_foreach_subgroup(base, subgroup_get_bis_indexes_cb, &bis_index);
+#endif /* CONFIG_TARGET_BROADCAST_CHANNEL */
 
-		return;
+	for (uint8_t subgroup = 0; subgroup < bis_index.num_subgroups; subgroup++) {
+		base_bis_index_bitfield |= bis_index.bis_index_bitfield[subgroup];
 	}
 
-#endif /* CONFIG_TARGET_BROADCAST_CHANNEL */
+	if (err != 0 || base_bis_index_bitfield == 0) {
+		printk("Failed to get valid BIS indexes: %d\n", err);
```

Comment:
If this fails on `base_bis_index_bitfield == 0` then printing the error code `0` won't be useful. 

Consider have 2 checks: One for `err != 0` and one for `base_bis_index_bitfield == 0` and more explicit print statements